File: tests/test_enrollment_receiver.py
#!/usr/bin/env python3
"""Exercise the real receiver and init command across lost-owner boundaries."""
import json
import os
from pathlib import Path
import signal
import subprocess
import tempfile
import time
import unittest


class EnrollmentReceiverTest(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory()
        self.addCleanup(self.tmp.cleanup)
        self.root = Path(self.tmp.name)
        self.repo = self.root / "project"
        subprocess.run(["git", "init", "-q", str(self.repo)], check=True)
        self.home = self.root / "home"
        self.home.mkdir()
        self.counter = self.root / "mutations"
        source = Path(__file__).resolve().parents[1]
        self.fleet = os.environ.get("HYDRA_FLEET_BIN", str(source / "build/hydra-fleet"))
        wrapper = self.root / "hydra-count"
        wrapper.write_text(
            "#!/usr/bin/env python3\n"
            "import os, pathlib, sys, time\n"
            "if 'init' in sys.argv:\n"
            "    pathlib.Path(os.environ['ENROLL_CHILD_PID']).write_text(str(os.getpid()))\n"
            "    pathlib.Path(os.environ['ENROLL_STARTED']).touch()\n"
            "    while os.environ.get('ENROLL_HOLD') and not pathlib.Path(os.environ['ENROLL_HOLD']).exists(): time.sleep(.01)\n"
            "    with open(os.environ['ENROLL_COUNTER'], 'a') as count: count.write('init\\n')\n"
            "os.execv(os.environ['ENROLL_REAL'], [os.environ['ENROLL_REAL'], *sys.argv[1:]])\n"
        )
        wrapper.chmod(0o755)
        self.env = dict(os.environ, HYDRA_HOME=str(self.home), HOME=str(self.root),
                        HYDRA_BIN_CMD=str(wrapper), ENROLL_REAL=str(source / "bin/hydra"),
                        ENROLL_COUNTER=str(self.counter), ENROLL_CHILD_PID=str(self.root / "child-pid"), ENROLL_STARTED=str(self.root / "started"))
        self.request = {"protocol": 1, "action": "init", "project": str(self.repo),
                        "args": ["--no-agent", "--trust"], "enrollment_operation_id": "a" * 64 + ":host", "expected_peer_fingerprint": "SHA256:reviewed"}

    def start(self, request=None):
        process = subprocess.Popen([self.fleet, "fleet", "serve"], env=self.env,
                                   stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                                   stderr=subprocess.PIPE, text=True, start_new_session=True)
        process.stdin.write(json.dumps(self.request if request is None else request))
        process.stdin.close()
        process.stdin = None
        return process

    def call(self, request=None):
        process = self.start(request)
        output, error = process.communicate(timeout=30)
        self.assertTrue(output, error)
        return json.loads(output)

    def count(self):
        return self.counter.read_text().splitlines() if self.counter.exists() else []

    def await_started(self, process):
        deadline = time.monotonic() + 10
        while time.monotonic() < deadline and process.poll() is None:
            if (self.root / "started").exists():
                return
            time.sleep(.01)
        self.fail("receiver did not reach the held init boundary")

    def kill_owner(self, process):
        if process.poll() is None:
            os.killpg(process.pid, signal.SIGKILL)
            child = self.root / "child-pid"
            if child.exists():
                try:
                    os.killpg(int(child.read_text()), signal.SIGKILL)
                except ProcessLookupError:
                    pass
        process.communicate(timeout=10)

    def test_duplicate_reconciles_exact_result_after_lost_response(self):
        first = self.call()
        self.assertTrue(first["ok"], first)
        # The caller may discard this response; retry must return its durable result.
        second = self.call()
        self.assertEqual(second, first)
        self.assertEqual(self.count(), ["init"])

    def test_changed_project_or_arguments_cannot_reuse_operation(self):
        self.assertTrue(self.call()["ok"])
        other = self.root / "other"
        subprocess.run(["git", "init", "-q", str(other)], check=True)
        for changed in (dict(self.request, project=str(other)), dict(self.request, args=["--no-agent"]), dict(self.request, expected_peer_fingerprint="SHA256:changed")):
            self.assertEqual(self.call(changed)["error"]["code"], "intent_changed")
        self.assertEqual(self.count(), ["init"])

    def test_concurrent_duplicate_has_one_mutation(self):
        release = self.root / "release"
        self.env["ENROLL_HOLD"] = str(release)
        first = self.start()
        self.addCleanup(self.kill_owner, first)
        self.await_started(first)
        self.assertEqual(self.call()["error"]["code"], "outcome_unknown")
        self.assertEqual(self.count(), [])
        release.touch()
        output, error = first.communicate(timeout=30)
        self.assertTrue(json.loads(output)["ok"], error)
        self.assertEqual(self.call(), json.loads(output))
        self.assertEqual(self.count(), ["init"])

    def test_owner_loss_preserves_unknown_without_replay(self):
        self.env["ENROLL_HOLD"] = str(self.root / "never-release")
        first = self.start()
        self.addCleanup(self.kill_owner, first)
        self.await_started(first)
        self.kill_owner(first)
        self.env.pop("ENROLL_HOLD")
        self.assertEqual(self.call()["error"]["code"], "outcome_unknown")
        self.assertEqual(self.count(), [])

    def test_invalid_ids_and_unavailable_state_have_zero_mutations(self):
        for operation in (None, 7, "../escape", "a" * 64 + ":../escape"):
            self.assertEqual(self.call(dict(self.request, enrollment_operation_id=operation))["error"]["code"], "invalid_input")
        (self.home / "fleet").mkdir()
        (self.home / "fleet/enrollment-ops").write_text("not a directory")
        self.assertEqual(self.call()["error"]["code"], "state_unavailable")
        self.assertEqual(self.count(), [])

    def test_legacy_or_corrupt_receipt_never_implies_success(self):
        directory = self.home / "fleet/enrollment-ops"
        directory.mkdir(parents=True)
        record = directory / ("a" * 64 + "_host")
        for value in ("succeeded", "pending", "{invalid"):
            record.write_text(value)
            self.assertEqual(self.call()["error"]["code"], "outcome_unknown")
        self.assertEqual(self.count(), [])


if __name__ == "__main__":
    unittest.main()

File: tests/native/test_enrollment_receiver.c
#include "support.h"
#include <errno.h>
#include <signal.h>
static char root[F_PATH], repo[F_PATH], home[F_PATH], counter[F_PATH],
    started[F_PATH], child_pid[F_PATH], fleet[F_PATH], fixture[F_PATH],
    hydra[F_PATH];
static json_object *request;
static void setup(void) {
  char tmp[] = "/tmp/hydra-native-enrollment-receiver-XXXXXX";
  assert(mkdtemp(tmp));
  assert(!f_copy(root, sizeof root, tmp));
  nt_path(repo, root, "project");
  char *args[] = {"git", "init", "-q", repo, NULL};
  struct f_capture c = nt_run(args);
  assert(!c.status);
  f_capture_free(&c);
  nt_path(home, root, "home");
  assert(!f_mkdirs(home));
  nt_path(counter, root, "mutations");
  nt_path(started, root, "started");
  nt_path(child_pid, root, "child-pid");
  assert(!setenv("HYDRA_HOME", home, 1) &&
         !setenv("HYDRA_BIN_CMD", fixture, 1) &&
         !setenv("ENROLL_REAL", hydra, 1) &&
         !setenv("ENROLL_COUNTER", counter, 1) &&
         !setenv("ENROLL_CHILD_PID", child_pid, 1) &&
         !setenv("ENROLL_STARTED", started, 1) && !unsetenv("ENROLL_HOLD"));
  request =
      f_parse("{\"protocol\":1,\"action\":\"init\",\"args\":[\"--no-agent\",\"-"
              "-trust\"],\"enrollment_operation_id\":"
              "\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
              "aa:host\",\"expected_peer_fingerprint\":\"SHA256:reviewed\"}");
  f_string_add(request, "project", repo);
}
static void cleanup(void) {
  json_object_put(request);
  assert(!f_remove_tree(root));
}
static struct nt_child start(json_object *r) {
  char *args[] = {fleet, "fleet", "serve", NULL};
  return nt_start(args, json_object_to_json_string_ext(r ? r : request,
                                                       JSON_C_TO_STRING_PLAIN));
}
static json_object *call(json_object *r) {
  struct nt_child child = start(r);
  struct f_capture c = nt_wait(&child, 30);
  if (!*c.out)
    fprintf(stderr, "%s\n", c.err);
  assert(!c.timeout && *c.out);
  json_object *v = f_parse(c.out);
  assert(v);
  f_capture_free(&c);
  return v;
}
static void error_is(json_object *v, const char *code) {
  assert(!strcmp(f_string(f_field(v, "error"), "code"), code));
  json_object_put(v);
}
static void count(bool one) {
  char *v = f_read(counter, 10000);
  if (one)
    assert(v && !strcmp(v, "init\n"));
  else
    assert(!v || !*v);
  free(v);
}
static void success(json_object *v) {
  nt_equal(f_field(v, "ok"), "true");
  json_object_put(v);
}
static void duplicate(void) {
  json_object *a = call(NULL), *b = call(NULL);
  nt_equal(f_field(a, "ok"), "true");
  assert(json_object_equal(a, b));
  json_object_put(a);
  json_object_put(b);
  count(true);
}
static void changed(void) {
  success(call(NULL));
  char other[F_PATH];
  nt_path(other, root, "other");
  char *args[] = {"git", "init", "-q", other, NULL};
  struct f_capture c = nt_run(args);
  assert(!c.status);
  f_capture_free(&c);
  for (size_t i = 0; i < 3; i++) {
    json_object *r = nt_clone(request);
    if (i == 0)
      f_string_add(r, "project", other);
    else if (i == 1)
      json_object_object_add(r, "args", f_parse_value("[\"--no-agent\"]"));
    else
      f_string_add(r, "expected_peer_fingerprint", "SHA256:changed");
    error_is(call(r), "intent_changed");
    json_object_put(r);
  }
  count(true);
}
static void concurrent(void) {
  char release[F_PATH];
  nt_path(release, root, "release");
  assert(!setenv("ENROLL_HOLD", release, 1));
  struct nt_child child = start(NULL);
  nt_await_file(started, 10);
  error_is(call(NULL), "outcome_unknown");
  count(false);
  nt_write(release, "");
  struct f_capture c = nt_wait(&child, 30);
  json_object *a = f_parse(c.out);
  assert(a);
  nt_equal(f_field(a, "ok"), "true");
  json_object *b = call(NULL);
  assert(json_object_equal(a, b));
  json_object_put(a);
  json_object_put(b);
  f_capture_free(&c);
  count(true);
}
static void owner_loss(void) {
  char release[F_PATH];
  nt_path(release, root, "never-release");
  assert(!setenv("ENROLL_HOLD", release, 1));
  struct nt_child child = start(NULL);
  nt_await_file(started, 10);
  assert(!kill(-child.pid, SIGKILL));
  char *pid = f_read(child_pid, 100);
  assert(pid);
  int rc = kill(-(pid_t)strtol(pid, NULL, 10), SIGKILL);
  assert(!rc || errno == ESRCH);
  free(pid);
  struct f_capture c = nt_wait(&child, 10);
  f_capture_free(&c);
  assert(!unsetenv("ENROLL_HOLD"));
  error_is(call(NULL), "outcome_unknown");
  count(false);
}
static void invalid(void) {
  const char *values[] = {"null", "7", "\"../escape\"",
                          ("\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
                          "aaaaaaaaaaaaaa:../escape\"")};
  for (size_t i = 0; i < 4; i++) {
    json_object *r = nt_clone(request);
    json_object_object_add(r, "enrollment_operation_id",
                           f_parse_value(values[i]));
    error_is(call(r), "invalid_input");
    json_object_put(r);
  }
  char dir[F_PATH], p[F_PATH];
  nt_path(dir, home, "fleet");
  assert(!f_mkdirs(dir));
  nt_path(p, dir, "enrollment-ops");
  nt_write(p, "not a directory");
  error_is(call(NULL), "state_unavailable");
  count(false);
}
static void corrupt(void) {
  char dir[F_PATH], p[F_PATH];
  nt_path(dir, home, "fleet/enrollment-ops");
  assert(!f_mkdirs(dir));
  nt_path(
      p, dir,
      "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_host");
  const char *values[] = {"succeeded", "pending", "{invalid"};
  for (size_t i = 0; i < 3; i++) {
    nt_write(p, values[i]);
    error_is(call(NULL), "outcome_unknown");
  }
  count(false);
}
int main(int argc, char **argv) {
  char origin[F_PATH];
  assert(getcwd(origin, sizeof origin));
  nt_path(hydra, origin, "bin/hydra");
  const char *f = getenv("HYDRA_FLEET_BIN");
  if (f)
    assert(!f_copy(fleet, sizeof fleet, f));
  else
    nt_path(fleet, origin, "build/hydra-fleet");
  if (argc > 1)
    assert(!f_copy(fixture, sizeof fixture, argv[1]));
  else
    nt_path(fixture, origin, "build/native-tests/enrollment-receiver-fixture");
  void (*cases[])(void) = {duplicate,  changed, concurrent,
                           owner_loss, invalid, corrupt};
  for (size_t i = 0; i < 6; i++) {
    setup();
    cases[i]();
    cleanup();
    printf("Enrollment receiver case %zu passed\n", i + 1);
  }
  return 0;
}
